File: blackjack-java/src/main/java/com/tobias/blackjack/service/GameService.java
package com.tobias.blackjack.service;

import com.tobias.blackjack.model.Game;
import com.tobias.blackjack.model.Player;
import org.springframework.stereotype.Service;

/**
 * Service class for managing the Blackjack game logic
 * 
 * This service handles the core game functionality and maintains
 * the current game state for web requests.
 */
@Service
public class GameService {

    private Game currentGame;

    /**
     * Start a new game with initial money
     */
    public Game startNewGame(int initialMoney) {
        Player player = new Player(initialMoney);
        currentGame = new Game(player);
        return currentGame;
    }

    /**
     * Player hits (takes another card)
     */
    public Game hit() {
        if (currentGame == null) {
            throw new IllegalStateException("No active game. Please start a new game first.");
        }
        
        if (!currentGame.isPlayerTurn()) {
            throw new IllegalStateException("It's not your turn to hit.");
        }
        
        currentGame.playerHit();
        return currentGame;
    }

    /**
     * Player stands (stops taking cards)
     */
    public Game stand() {
        if (currentGame == null) {
            throw new IllegalStateException("No active game. Please start a new game first.");
        }
        
        if (!currentGame.isPlayerTurn()) {
            throw new IllegalStateException("It's not your turn to stand.");
        }
        
        currentGame.playerStand();
        return currentGame;
    }

    /**
     * Place a bet and deal initial cards
     */
    public Game placeBet(int amount) {
        if (currentGame == null) {
            throw new IllegalStateException("No active game. Please start a new game first.");
        }
        
        // Check if we can place a bet (game must be finished, new, or cards already dealt)
        String status = currentGame.getGameStatus();
        if (!status.equals("NEW_GAME") && 
            !status.equals("PLAYER_WINS") && 
            !status.equals("DEALER_WINS") && 
            !status.equals("PLAYER_BUSTED") && 
            !status.equals("DEALER_BUSTED") && 
            !status.equals("PUSH") &&
            !status.equals("CARDS_DEALT")) {
            throw new IllegalStateException("Cannot place bet while game is in progress.");
        }
        
        currentGame.placeBet(amount);
        // Only deal new cards if status is not CARDS_DEALT (cards already dealt)
        if (!status.equals("CARDS_DEALT")) {
            currentGame.dealInitialCards();
        }
        return currentGame;
    }

    /**
     * Get the current game state
     */
    public Game getCurrentGame() {
        if (currentGame == null) {
            throw new IllegalStateException("No active game. Please start a new game first.");
        }
        return currentGame;
    }

    /**
     * Check if there's an active game
     */
    public boolean hasActiveGame() {
        return currentGame != null;
    }

    /**
     * Deal cards without starting the game (for "Play Again" functionality)
     */
    public Game dealCards() {
        if (currentGame == null) {
            throw new IllegalStateException("No active game. Please start a new game first.");
        }
        
        // Check if we can deal cards (game must be finished or new)
        String status = currentGame.getGameStatus();
        if (!status.equals("NEW_GAME") && 
            !status.equals("PLAYER_WINS") && 
            !status.equals("DEALER_WINS") && 
            !status.equals("PLAYER_BUSTED") && 
            !status.equals("DEALER_BUSTED") && 
            !status.equals("PUSH") &&
            !status.equals("CARDS_DEALT")) {
            throw new IllegalStateException("Cannot deal cards while game is in progress.");
        }
        
        currentGame.dealCardsHidden();
        return currentGame;
    }

    /**
     * Start the game by revealing cards and checking for blackjack
     */
    public Game startGame() {
        if (currentGame == null) {
            throw new IllegalStateException("No active game. Please start a new game first.");
        }
        
        // Only start if cards are already dealt
        if (!currentGame.getGameStatus().equals("CARDS_DEALT")) {
            throw new IllegalStateException("Cards must be dealt first before starting the game.");
        }
        
        // Check if bet is placed
        if (currentGame.getPlayer().getCurrentBet() <= 0) {
            throw new IllegalStateException("You must place a bet before starting the game.");
        }
        
        // Change status to PLAYING and check for blackjack
        currentGame.setGameStatus("PLAYING");
        
        // Check for blackjack
        if (currentGame.getPlayer().hasBlackjack() && currentGame.getDealerValue() == 21) {
            // Both have blackjack - push
            currentGame.setPlayerTurn(false);
            currentGame.setGameStatus("PUSH");
            // Return bet
            int currentBet = currentGame.getPlayer().getCurrentBet();
            currentGame.getPlayer().setMoney(currentGame.getPlayer().getMoney() + currentBet);
            currentGame.getPlayer().setCurrentBet(0);
        } else if (currentGame.getPlayer().hasBlackjack()) {
            // Player has blackjack - instant win (pays 3:2)
            currentGame.setPlayerTurn(false);
            currentGame.setGameStatus("PLAYER_WINS");
            int bet = currentGame.getPlayer().getCurrentBet();
            // Blackjack pays 3:2, so winnings = bet * 1.5
            int winnings = (bet * 3) / 2; // 3:2 odds
            currentGame.getPlayer().setMoney(currentGame.getPlayer().getMoney() + bet + winnings);
            currentGame.getPlayer().setCurrentBet(0);
        } else if (currentGame.getDealerValue() == 21) {
            // Dealer has blackjack - instant loss
            currentGame.setPlayerTurn(false);
            currentGame.setGameStatus("DEALER_WINS");
            // Bet already deducted, just clear it
            currentGame.getPlayer().setCurrentBet(0);
        }
        
        return currentGame;
    }
}

File: src/lib/blackjack/GameService.ts
import { Game } from './Game';
import { Player } from './Player';

// Store games in memory (in production, you might want to use a database or Redis)
const games = new Map<string, Game>();

export class GameService {
  private static getGameId(sessionId?: string): string {
    // In a real app, you'd use a proper session ID
    // For now, we'll use a default session or generate one
    return sessionId || 'default';
  }

  static startNewGame(initialMoney: number = 100, sessionId?: string): Game {
    const player = new Player(initialMoney);
    const game = new Game(player);
    const gameId = this.getGameId(sessionId);
    games.set(gameId, game);
    return game;
  }

  static getCurrentGame(sessionId?: string): Game {
    const gameId = this.getGameId(sessionId);
    const game = games.get(gameId);
    if (!game) {
      throw new Error('No active game. Please start a new game first.');
    }
    return game;
  }

  static hit(sessionId?: string): Game {
    const game = this.getCurrentGame(sessionId);
    
    if (!game.isPlayerTurn) {
      throw new Error("It's not your turn to hit.");
    }
    
    game.playerHit();
    return game;
  }

  static stand(sessionId?: string): Game {
    const game = this.getCurrentGame(sessionId);
    
    if (!game.isPlayerTurn) {
      throw new Error("It's not your turn to stand.");
    }
    
    game.playerStand();
    return game;
  }

  static placeBet(amount: number, sessionId?: string): Game {
    const game = this.getCurrentGame(sessionId);
    
    // Check if we can place a bet
    const status = game.gameStatus;
    if (status !== 'NEW_GAME' && 
        status !== 'PLAYER_WINS' && 
        status !== 'DEALER_WINS' && 
        status !== 'PLAYER_BUSTED' && 
        status !== 'DEALER_BUSTED' && 
        status !== 'PUSH' &&
        status !== 'CARDS_DEALT') {
      throw new Error('Cannot place bet while game is in progress.');
    }
    
    game.placeBet(amount);
    // Only deal new cards if status is not CARDS_DEALT
    if (status !== 'CARDS_DEALT') {
      game.dealInitialCards();
    }
    return game;
  }

  static dealCards(sessionId?: string): Game {
    const game = this.getCurrentGame(sessionId);
    
    // Check if we can deal cards
    const status = game.gameStatus;
    if (status !== 'NEW_GAME' && 
        status !== 'PLAYER_WINS' && 
        status !== 'DEALER_WINS' && 
        status !== 'PLAYER_BUSTED' && 
        status !== 'DEALER_BUSTED' && 
        status !== 'PUSH' &&
        status !== 'CARDS_DEALT') {
      throw new Error('Cannot deal cards while game is in progress.');
    }
    
    game.dealCardsHidden();
    return game;
  }

  static startGame(sessionId?: string): Game {
    const game = this.getCurrentGame(sessionId);
    
    // Only start if cards are already dealt
    if (game.gameStatus !== 'CARDS_DEALT') {
      throw new Error('Cards must be dealt first before starting the game.');
    }
    
    // Check if bet is placed
    if (game.player.currentBet <= 0) {
      throw new Error('You must place a bet before starting the game.');
    }
    
    game.startGame();
    return game;
  }

  static hasActiveGame(sessionId?: string): boolean {
    const gameId = this.getGameId(sessionId);
    return games.has(gameId);
  }
}

